crates/rustc_codegen_spirv/src/builder/builder_methods.rs

```diff
@@ -540,190 +578,434 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
+                // Original access chain had no indices to merge with.
+                false
+            };
+
+            if can_merge {
+                let last_original_idx_id = original_indices.last_mut().unwrap();
+                // Add the current `ptr_base_index` to the last index of the original chain.
+                // The result becomes the new last index.
+                *last_original_idx_id = self
+                    .add(
+                        // Ensure types match for add.
+                        last_original_idx_id.with_type(ptr_base_index.ty),
+                        ptr_base_index,
+                    )
+                    // Define the result of the addition.
+                    .def(self);
+                // Append the remaining indices (`indices`) from the current GEP operation.
+                original_indices.extend(gep_indices_ids);
+
+                trace!(
+                    "emitting merged access chain with pointer to type: {}",
+                    self.debug_type(calculated_pointee_type)
+                );
+                // Emit a *single* AccessChain using the *original* base pointer and the *merged* index list.
+                // The result type *must* be the `final_spirv_ptr_type` calculated earlier based on the full chain of operations.
+                return self.emit_access_chain(
+                    final_spirv_ptr_type, // Use the strictly calculated final type.
+                    original_ptr,         // Base pointer from the *original* AccessChain.
+                    None,                 // No separate base index; it's merged.
+                    original_indices,     // The combined list of indices.
+                    is_inbounds,          // Preserve original inbounds request.
+                );
+            } else {
+                // Cannot merge because one or both relevant indices are not constant,
+                // or the original chain was empty.
+                trace!(
+                    "Last index or base offset is not constant, or no last index, cannot merge."
+                );
+            }
+        } else {
+            // The base pointer `ptr` was not the result of an AccessChain, or merging
+            // wasn't attempted due to type mismatch.
+            trace!("no original access chain to merge with");
         }
+        // --- End GEP Merging Path ---
 
-        // HACK(eddyb) temporary workaround for untyped pointers upstream.
-        // FIXME(eddyb) replace with untyped memory SPIR-V + `qptr` or similar.
+        // --- Fallback / Default Path ---
+
+        // This path is taken if neither the Recovery nor the Merging path succeeded or applied.
+        // It performs a more direct translation of the GEP request.
+
+        // HACK(eddyb): Workaround for potential upstream issues where pointers might lack precise type info.
+        // FIXME(eddyb): Ideally, this should use untyped memory features if available/necessary.
```

Comment:
These should've stayed next to the `pointercast` - the new comments might not even make sense, because the cast is only due to the mismatch - anyway, it's all a historical mess, and lots of concepts don't fit the new (much more untyped) approach, sadly.